### PR TITLE
Scenario inheritance

### PR DESCRIPTION
**Pull request by knoxsp**
The idea with this new feature is to ensure existing functionality remains unchanged. 
The new concept of a child scenario only affects functions which request a scenario or scenario-relevant data. In order to get the inherited data, you must use the flag get_parent_data=True.

### CODE REVIEW

File: tests/test_network.py [hunk 1]
```diff
@@ -631,24 +631,6 @@ def test_set_link_status(self, session, network_with_data):
         assert link_to_delete.id in link_ids
 
 
-    def test_delete_link(self, session, network_with_data):
```
Comment (jetuk) [inline tests/test_network.py]: Why was this removed?

File: tests/test_scenario.py [hunk 2]
```diff
@@ -548,6 +548,136 @@ def test_clone(self, session, network_with_data):
 
         return updated_network
 
+
+    def test_scenario_inheritance(self, session, network_with_data):
```
Comment (jetuk) [inline tests/test_scenario.py]: Maybe split this into multiple tests for clarity?

Comment (knoxsp) [reply tests/test_scenario.py]: Done

File: hydra_base/db/alembic.ini [hunk 3]
```diff
@@ -35,7 +35,7 @@ script_location = alembic
 # are written from script.py.mako
 # output_encoding = utf-8
 
-sqlalchemy.url = mysql+mysqlconnector://root:root@localhost/hydradb
+sqlalchemy.url = mysql+mysqldb://root:@localhost/hydradb
```
Comment (jetuk) [inline hydra_base/db/alembic.ini]: Is this change needed in here?

File: hydra_base/db/model.py [hunk 4]
```diff
@@ -1298,6 +1300,65 @@ def add_resourcegroup_item(self, ref_key, resource, group_id):
...
+        for child_rs in child_data:
+            childrens_ras.append(child_rs.resource_attr_id)
+        
+        #Add resource attributes which are not defined already 
+        for this_rs in self.resourcescenarios:
+            if this_rs.resource_attr_id not in childrens_ras:
+                child_data.append(this_rs)
+
+        if self.parent is not None and get_parent_data is True:
+            return self.parent.get_data(child_data)
```
Comment (jetuk) [inline hydra_base/db/model.py]: Does this need to pass on `get_parent_data=get_parent_data`. Currently it will only do 1 recursion?

File: hydra_base/db/model.py [hunk 5]
```diff
@@ -1298,6 +1300,65 @@ def add_resourcegroup_item(self, ref_key, resource, group_id):
...
+            child_data = []
+
+        #Idenify all existing resource attr ids, which take priority over anything in here
+        childrens_ras = []
+        for child_rs in child_data:
+            childrens_ras.append(child_rs.resource_attr_id)
+        
+        #Add resource attributes which are not defined already 
+        for this_rs in self.resourcescenarios:
+            if this_rs.resource_attr_id not in childrens_ras:
```
Comment (jetuk) [inline hydra_base/db/model.py]: I guess this is not going to work in the case where you delete some data in a child scenario? It'll revert to using previous undeleted data? What's the user intended to do in this situation?

File: hydra_base/db/model.py [hunk 6]
```diff
@@ -1298,6 +1300,65 @@ def add_resourcegroup_item(self, ref_key, resource, group_id):
...
+        for child_rgi in child_items:
+            childrens_groups.append(child_rgi.group_id)
+        
+        #Add resource attributes which are not defined already 
+        for this_rgi in self.resourcegroupitems:
+            if this_rgi.group_id not in childrens_groups:
+                child_items.append(this_rgi)
+
+        if self.parent is not None and get_parent_items is True:
+            return self.parent.get_group_items(child_items)
```
Comment (jetuk) [inline hydra_base/db/model.py]: Does this need to pass on `get_parent_data=get_parent_data`. Currently it will only do 1 recursion?

File: hydra_base/lib/groups.py [hunk 7]
```diff
@@ -53,11 +53,14 @@ def add_resourcegroup(group, network_id,**kwargs):
...
 
-def delete_resourcegroup(group_id,**kwargs):
+def delete_resourcegroup(group_id, purge_data='N', **kwargs):
     """
         Add a new group to a scenario.
     """
     group_i = _get_group(group_id)
+
+    if purge_data == 'Y':
+        network._purge_datasets_unique_to_resource('GROUP', group_id)
```
Comment (jetuk) [inline hydra_base/lib/groups.py]: Has this got anything to do with scenario inheritance?

File: hydra_base/lib/scenario.py [hunk 8]
```diff
@@ -311,6 +313,59 @@ def purge_scenario(scenario_id, **kwargs):
     db.DBSession.flush()
     return 'OK'
 
+def create_child_scenario(scenario_id, child_name, **kwargs):
```
Comment (jetuk) [inline hydra_base/lib/scenario.py]: Maybe call `scenario_id` -> `parent_scenario_id` to be more explicit?

File: hydra_base/lib/scenario.py [hunk 9]
```diff
@@ -122,7 +122,7 @@ def copy_data_from_scenario(resource_attrs, source_scenario_id, target_scenario_
 
     return target_resourcescenarios
 
-def get_scenario(scenario_id,**kwargs):
+def get_scenario(scenario_id, get_parent_data=False, **kwargs):
```
Comment (jetuk) [inline hydra_base/lib/scenario.py]: Should `get_parent_data` default to `True` for scenarios with a parent?

File: tests/test_scenario.py [hunk 10]
```diff
@@ -548,6 +548,149 @@ def test_clone(self, session, network_with_data):
 
         return updated_network
 
+    def test_get_inherited_data(self, session, network_with_child_scenario):
```
Comment (jetuk) [inline tests/test_scenario.py]: Can we add a test with a grand-parent to test recursion past the first level?

File: hydra_base/db/model.py (hunk 5 above)
Comment (knoxsp) [reply hydra_base/db/model.py]: I think this is a UI issue, rathar than being the responsibility of HB

File: hydra_base/lib/groups.py (hunk 7 above)
Comment (knoxsp) [reply hydra_base/lib/groups.py]: no, it was just a fix I noticed :)

File: hydra_base/lib/scenario.py (hunk 9 above)
Comment (knoxsp) [reply hydra_base/lib/scenario.py]: By making it false, the rational was to make retrieval of data not directly inside a scenario an explicit operation. This way i hope it will reduce misunderstanding of what the function is expected to return

File: hydra_base/lib/scenario.py (hunk 8 above)
Comment (knoxsp) [reply hydra_base/lib/scenario.py]: 026aa448918f3857932bd386ead95eb43441499d

File: hydra_base/db/alembic.ini (hunk 3 above)
Comment (knoxsp) [reply hydra_base/db/alembic.ini]: Fixed with merge from master

File: hydra_base/db/model.py (hunk 4 above)
Comment (knoxsp) [reply hydra_base/db/model.py]: done, with test included

File: hydra_base/db/model.py (hunk 6 above)
Comment (knoxsp) [reply hydra_base/db/model.py]: This did work because the get_parent_items defaulted to true. I've defaulted it to false, in line with the other function, and now I pass it in explicitly as recommended. Tests included.

File: tests/test_scenario.py (hunk 10 above)
Comment (knoxsp) [reply tests/test_scenario.py]: Done

**Comment by jetuk** [comment]
Any idea why the test  fails on postgres against master?
